### PR TITLE
[ML] Address some bottlenecks training classification and regression models on large data sets

### PR DESCRIPTION
**Pull request by tveasey**
Investigating slowdowns on a couple of large QA data sets as a result of #2298 showed up some interesting things. We spend a lot of time updating quantiles and recomputing split indices for the data sets which slowed down. These data sets have relatively many metric features and the runtime share of these operations is larger in such cases. This change introduces some changes to address long run times in such cases. 

I'd planned to use a fast version of the quantile sketch, but a silly mistake meant it never actually got wired in. I've also made a further tweak to the fast version which gives it another 50% performance boost at the expense of slightly changing its behaviour. I tested accuracy fairly extensively for the new scheme, which still often produces the same results, and found no evidence for significantly different accuracy.

For many features we can gain a lot of performance with v.little impact on the QoR by updating candidate splits less frequently. This imposes a new limit on the frequency we fresh them based on feature count.

For large data sets we were allowing individual trees to become v.large. Such large trees both stresses our inference, we have to limit the model size based on its memory anyway, and means train time can become too high. I've capped the maximum tree size independent of the data set size. The limit is still pretty large and I measured no significant impact in QoR from this change.

### CODE REVIEW

File: lib/maths/common/unittest/CBasicStatisticsTest.cc [hunk 1]
```diff
@@ -1121,63 +1121,14 @@ BOOST_AUTO_TEST_CASE(testCovariancesLedoitWolf) {
...
-        double sample[] = {3.0, 5.0, 10.0, 2.0, 6.0, 7.0, 1.0, 9.0, 4.0, 8.0};
-
-        maths::common::CBasicStatistics::TDoubleVec sampleVec(
-            sample, sample + sizeof(sample) / sizeof(sample[0]));
-
-        double median = maths::common::CBasicStatistics::median(sampleVec);
-
-        BOOST_REQUIRE_EQUAL(5.5, median);
-    }
+    BOOST_REQUIRE_EQUAL(0.0, maths::common::CBasicStatistics::median({}));
```
Comment (valeriy42) [inline lib/maths/common/unittest/CBasicStatisticsTest.cc:1124]: nice simplification!

File: lib/maths/common/CQuantileSketch.cc [hunk 2]
```diff
@@ -571,49 +588,68 @@ void CFastQuantileSketch::debugMemoryUsage(const core::CMemoryUsage::TMemoryUsag
...
 
-        this->reduce(m_Rng, m_MergeCosts, m_MergeCandidates, m_Stale);
-    }
-}
+        std::size_t numberToMerge{knots.size() - this->fastReduceTargetSize()};
 
-std::size_t CFastQuantileSketch::target() const {
-    return static_cast<std::size_t>(
-        m_ReductionFraction * static_cast<double>(this->maxSize()) + 1.0);
+        // This is pseudo greedy since unlike CQuantileSketch which is greedy.
```
Comment (valeriy42) [inline lib/maths/common/CQuantileSketch.cc:630]:
```suggestion
        // This is pseudo greedy unlike CQuantileSketch which is greedy.
```

File: lib/maths/analytics/unittest/CBoostedTreeTest.cc [hunk 3]
```diff
@@ -1618,12 +1620,8 @@ BOOST_AUTO_TEST_CASE(testMseIncrementalAddNewTrees) {
               << ", 5 new trees = " << testError5 << ", 10 new trees = " << testError10);
     // The initial model has too little capacity so we should get substantial
     // improvements for adding trees.
-    BOOST_TEST_REQUIRE(0.95 * testError0 >= testError5);
+    BOOST_TEST_REQUIRE(0.9 * testError0 >= testError5);
     BOOST_TEST_REQUIRE(0.9 * testError0 >= testError10);
```
Comment (valeriy42) [inline lib/maths/analytics/unittest/CBoostedTreeTest.cc:1624]: I guess, in this form we could remove `testError10` altogether to simplify.

Comment (tveasey) [reply lib/maths/analytics/unittest/CBoostedTreeTest.cc:1624]: Actually good point. I think 5 vs 10 is just a fragile. This has failed for more under perturbations before. You have to arrange for the necessary capacity to be just right. Perhaps if we overrode more hyperparameter values we could make it more reliable, but I think 0 vs 5 is sufficient if you're happy?

File: lib/maths/common/CQuantileSketch.cc (hunk 2 above)
Comment (tveasey) [reply lib/maths/common/CQuantileSketch.cc]: See [3671abb](https://github.com/elastic/ml-cpp/pull/2332/commits/3671abbe45844fe59591aeeb40ad435bf93c27ec).

**Review by valeriy42 — APPROVED** [review]
LGTM

**Pull request merged**